bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -0,0 +1,321 @@
+package jsky.app.ot.ags
+
+import java.beans.{PropertyChangeEvent, PropertyChangeListener}
+import java.util.concurrent.{LinkedBlockingQueue, Executors}
+import java.util.logging.Logger
+
+import edu.gemini.ags.api.{AgsRegistrar, AgsStrategy}
+import edu.gemini.pot.sp._
+import edu.gemini.spModel.guide.GuideProbe
+import edu.gemini.spModel.obs.context.ObsContext
+import edu.gemini.spModel.rich.shared.immutable._
+import edu.gemini.spModel.target.env.{TargetEnvironment, GuideProbeTargets}
+import edu.gemini.spModel.target.obsComp.TargetObsComp
+import jsky.app.ot.OT
+import jsky.app.ot.gemini.altair.Altair_WFS_Feature
+import jsky.app.ot.gemini.inst.OIWFS_Feature
+import jsky.app.ot.gemini.tpe.TpePWFSFeature
+import jsky.app.ot.tpe.{GemsGuideStarWorker, GuideStarSupport, TpeContext, TpeManager}
+import scala.collection.JavaConverters._
+import scala.collection.immutable.HashMap
+import scala.swing.Swing
+import scala.util.{Try, Failure, Success}
+import scala.concurrent.ExecutionContext.Implicits.global
+
+
+object BagsManager {
+  private val LOG: Logger = Logger.getLogger(getClass.getName)
+
+  private case class BagsTask(observation: ISPObservation, curCtx: Option[ObsContext], nextCtx: Option[ObsContext]) {
+    def obsKey: SPNodeKey =
+      observation.getNodeKey
+
+    // The state to move to when processing the task.
+    def nextTask: Option[BagsTask] =
+      nextCtx.map(_ => BagsTask(observation, nextCtx, None))
+
+    def isDone: Boolean =
+      nextCtx.isEmpty
+
+    def newContext(ctxOpt: Option[ObsContext]): BagsTask =
+      BagsTask(observation, curCtx, ctxOpt)
+
+    def performLookup(): Unit = {
+      LOG.info(s"Performing BAGS lookup for observation=${observation.getObservationID}")
+
+
+      def resultString[A](r: Option[A]): String =
+        s"Results=${r.fold("No")(_ => "Yes")}."
+
+      // We must handle GeMS separately from other strategies, and delegate the work to the GemsGuideStarWorker.
+      def gemsLookup(obsCtx: ObsContext): Unit = {
+        LOG.info(s"BAGS GeMS lookup for observation=${observation.getObservationID} starting...")
+        val ggsTry = Try { GemsGuideStarWorker.findGuideStars(obsCtx) }
```

Comment:
@tpolecat Are we rather using `\/` than `Try`? In any case you could do patter matching on the try to make it a bit cleaner
```
ggsTry match {
    case Failure(x) => LOG.warning("....")
    case Sucess(gss) => Swing.onEDT {...
}
```